data/finset/basic.lean

```diff
@@ -0,0 +1,704 @@
+     | decidable.is_false n := decidable.is_false (λ e : ⟦l₁⟧ = ⟦l₂⟧, absurd (quotient.exact e) n)
+end)
+
+def mem (a : α) (s : finset α) : Prop :=
+quot.lift_on s (λ l, a ∈ l.1)
+ (λ l₁ l₂ (e : l₁ ~ l₂), propext (iff.intro
+   (λ ainl₁, perm.mem_of_perm e ainl₁)
+   (λ ainl₂, perm.mem_of_perm (perm.symm e) ainl₂)))
+
+-- instance : has_mem α (finset α) := ⟨mem⟩  -- this causes a failure of type class inference
+infix ∈ := mem
+notation a ∉ b := ¬ mem a b
+
+theorem mem_of_mem_list {a : α} {l : nodup_list α} : a ∈ l.1 → a ∈ ⟦l⟧ :=
+λ ainl, ainl
+
+theorem mem_list_of_mem {a : α} {l : nodup_list α} : a ∈ ⟦l⟧ → a ∈ l.1 :=
+λ ainl, ainl
+
+instance decidable_mem [h : decidable_eq α] : ∀ (a : α) (s : finset α), decidable (a ∈ s) :=
+λ a s, quot.rec_on_subsingleton s
+  (λ l, match list.decidable_mem a l.1 with
+        | decidable.is_true p := decidable.is_true (mem_of_mem_list p)
+        | decidable.is_false n := decidable.is_false (λ p, absurd (mem_list_of_mem p) n)
+        end)
+
+theorem mem_to_finset [decidable_eq α] {a : α} {l : list α} : a ∈ l → a ∈ to_finset l :=
+λ ainl, mem_erase_dup.mpr ainl
+
+theorem mem_to_finset_of_nodup {a : α} {l : list α} (n : nodup l) : a ∈ l → a ∈ to_finset_of_nodup l n :=
+λ ainl, ainl
+
+/- extensionality -/
+theorem ext {s₁ s₂ : finset α} : (∀ a, a ∈ s₁ ↔ a ∈ s₂) → s₁ = s₂ :=
+quotient.induction_on₂ s₁ s₂ (λ l₁ l₂ e, quot.sound (perm.perm_ext l₁.2 l₂.2 e))
+
+/- empty -/
+def empty : finset α :=
+to_finset_of_nodup [] nodup_nil
+
+instance : has_emptyc (finset α) := ⟨empty⟩
+
+attribute [simp]
+theorem not_mem_empty (a : α) : a ∉ (∅ : finset α) :=
+λ aine, aine
+
+attribute [simp]
+theorem mem_empty_iff (x : α) : x ∈ (∅ : finset α) ↔ false :=
+iff_false_intro (not_mem_empty _)
+
+theorem mem_empty_eq (x : α) : x ∈ (∅ : finset α) = false :=
+propext (mem_empty_iff _)
+
+theorem eq_empty_of_forall_not_mem {s : finset α} (H : ∀x, x ∉ s) : s = ∅ :=
+ext (λ x, iff_false_intro (H x))
+
+-- /- universe -/
+-- def univ [h : fintype A] : finset A :=
+-- to_finset_of_nodup (@fintype.elems A h) (@fintype.unique A h)
+
+-- theorem mem_univ [fintype A] (x : A) : x ∈ univ :=
+-- fintype.complete x
+
+-- theorem mem_univ_eq [fintype A] (x : A) : x ∈ univ = true := propext (iff_true_intro !mem_univ)
+
+/- card -/
+def card (s : finset α) : nat :=
+quot.lift_on s
+  (λ l, length l.1)
+  (λ l₁ l₂ p, perm.length_eq_length_of_perm p)
+
+theorem card_empty : card (@empty α) = 0 :=
+rfl
+
+lemma ne_empty_of_card_eq_succ {s : finset α} {n : nat} : card s = succ n → s ≠ ∅ :=
+λ h hn, by rw hn at h; contradiction
+
+/- insert -/
+section insert
+variable [h : decidable_eq α]
+include h
+
+def insert (a : α) (s : finset α) : finset α :=
+quot.lift_on s
+  (λ l, to_finset_of_nodup (insert a l.1) (nodup_insert l.2))
+  (λ (l₁ l₂ : nodup_list α) (p : l₁ ~ l₂), quot.sound (perm.perm_insert a p))
+
+-- set builder notation
+notation `'{`:max a:(foldr `, ` (x b, insert x b) ∅) `}`:0 := a
```

Comment:
I guess if you use the `has_insert` type class then the set builder notation is not necessary and `singleton` is already defined.